src/client/service/identitiesInfoConnectedGet.ts

```diff
@@ -0,0 +1,103 @@
+import type * as grpc from '@grpc/grpc-js';
+import type { Authenticate } from '../types';
+import type { IdentitiesManager } from '../../identities';
+import type { IdentityData, ProviderId } from '../../identities/types';
+import { validateSync } from '../../validation';
+import { matchSync } from '../../utils';
+import * as grpcUtils from '../../grpc/utils';
+import * as validationUtils from '../../validation/utils';
+import * as identitiesErrors from '../../identities/errors';
+import * as identitiesPB from '../../proto/js/polykey/v1/identities/identities_pb';
+
+function identitiesInfoConnectedGet({
+  identitiesManager,
+  authenticate,
+}: {
+  identitiesManager: IdentitiesManager;
+  authenticate: Authenticate;
+}) {
+  return async (
+    call: grpc.ServerWritableStream<
+      identitiesPB.ProviderSearch,
+      identitiesPB.Info
+    >,
+  ): Promise<void> => {
+    const genWritable = grpcUtils.generatorWritable(call);
+    try {
+      const metadata = await authenticate(call.metadata);
+      call.sendMetadata(metadata);
+      const {
+        providerIds,
+      }: {
+        providerIds: Array<ProviderId>;
+      } = validateSync(
+        (keyPath, value) => {
+          return matchSync(keyPath)(
+            [['providerIds'], () => value.map(validationUtils.parseProviderId)],
+            () => value,
+          );
+        },
+        {
+          providerIds: call.request.getProviderIdList(),
+        },
+      );
+      // Process options that were set
+      if (providerIds.length === 0) {
+        Object.keys(identitiesManager.getProviders()).forEach((id) =>
+          providerIds.push(id as ProviderId),
+        );
+      }
+      const getDisconnected = call.request.getDisconnected();
+      if (getDisconnected) {
+        // Can only get connected identities at this stage
+        throw new identitiesErrors.ErrorProviderUnimplemented();
+      }
+      const identities: Array<AsyncGenerator<IdentityData, any, unknown>> = [];
+      for (const providerId of providerIds) {
+        // Get provider from id
+        const provider = identitiesManager.getProvider(providerId);
+        if (provider === undefined)
+          throw new identitiesErrors.ErrorProviderMissing();
+        // Get our own authenticated identity in order to query, skip provider
+        // if not authenticated
+        const authIdentities = await provider.getAuthIdentityIds();
+        if (authIdentities.length === 0) {
+          break;
+        }
+        identities.push(
+          provider.getConnectedIdentityDatas(
+            authIdentities[0],
+            call.request.getSearchTermList(),
+          ),
+        );
```

Comment:
`authIdentities[0]` assumes that we want to use the first authenticated identity to perform this search - but will this always be the case? What if I have multiple GitHub accounts? Would it be better to use all of our authenticated identities on some provider to inform the search?

Reply:
Yeah that's a good point... previously the `authIdentityId` field wasn't being used correctly in most places where the provider was used (i.e. it was being treated as the same thing as `identityId`) so I just went with the easiest option to implement. Not sure if a better solution would be to iterate through all of your authenticated ids every time or whether setting a specifc one to use should be an argument of the `identities search` command (either optional or required)? @CMCDragonkai do you have a preference here?

Reply:
So we can default to the first auth identity id, and allow the user to select a different one in the parameters.

Is there a separate call that you can do to know what auth identity ids you have?

Reply:
There's a method on the provider, so I'd just need to create a relevant command + handler to do it. For the naming maybe something like `identitiesAuthenticatedGet` and `identities authenticated`? Might be a bit too similar to `identities authenticate` though.

Reply:
> So we can default to the first auth identity id, and allow the user to select a different one in the parameters.

There is actually one problem with doing this that I just realised - with our new validation system an identity id that is an empty string is considered invalid, however, if we don't set a parameter then the default proto value is an empty string. So this means the expected behaviour of an empty string being used to represent a value not being set will be considered a validation error. A simple solution would be just not validating the identity id, but this could cause problems in the future if we ever change the way that identity ids are validated.

Reply:
The alternative would be to create an extra field in the `ProviderSearch` message to check for whether the auth identity id is set or not, and only validate it if this boolean is set, e.g.
```proto
message ProviderSearch {
  string auth_identity_id = 1;
  bool auth_identity_id_set = 2; // set this to true if the value above has been explicitly set
  string identity_id = 3;
  bool disconnected = 4;
  string limit = 5;
  repeated string search_term = 6;
  repeated string provider_id = 7;
}
```

Reply:
The auth identities are stored inside `IdentitiesManager`, so we can't get the first auth identity from the client side (it has to be done on the server side). Becaue of this, a value might not be set for that field, in which case it would be an empty string.

Reply:
Try using:

```
oneof identity_id {
  string string_value = 1;
}
```

If the `oneof` results in a potentially `undefined` value in the JS, then that can be your sentinel value.

Reply:
We basically need an "explicit" `optional` for proto3. And I'm wondering if `oneof` is what we need for that situation: https://developers.google.com/protocol-buffers/docs/proto3#oneof

Otherwise `NullValue` from `struct.proto` could be used as that sentinel case.

Reply:
According to this answer https://stackoverflow.com/a/42634681/582917, you can use `oneof` to represent this explicit optional. Also future versions of proto3 will reintroduce the explicit `optional` keyword that we can use. At least for now, the `oneof` is more semantically correct to represent an explicitly optional property of a message.

Reply:
We're still in proto3 3.14. The next version of proto3 will have the explicit `optional` available. So we will use `oneof` for now.

Reply:
Note that the name of the `oneof`. I'm not sure what the effect of each name will be. https://stackoverflow.com/a/46821842/582917

This answer suggests doing:

```
oneof optional_identity_id {
  string identity_id = 1;
}
```

You can try it out and see how it goes.

Reply:
Using
```proto
oneof optional_auth_identity_id {
  string auth_identity_id = 1;
}
```
Generates the following
```ts
hasAuthIdentityId(): boolean;
clearAuthIdentityId(): void;
getAuthIdentityId(): string;
getOptionalAuthIdentityIdCase(): ProviderSearch.OptionalAuthIdentityIdCase;

export namespace ProviderSearch {
    export type AsObject = {
        authIdentityId: string,
        identityId: string,
        disconnected: boolean,
        limit: string,
        searchTermList: Array<string>,
        providerIdList: Array<string>,
    }

    export enum OptionalAuthIdentityIdCase {
        OPTIONAL_AUTH_IDENTITY_ID_NOT_SET = 0,
        AUTH_IDENTITY_ID = 1,
    }

}
```

So we can call `hasAuthIdentityId()` to check if it has been explicitly set, and if so we can validate the value of `getAuthIdentityId()`.

Reply:
> There's a method on the provider, so I'd just need to create a relevant command + handler to do it. For the naming maybe something like `identitiesAuthenticatedGet` and `identities authenticated`? Might be a bit too similar to `identities authenticate` though.

I've created a CLI command and rpc handler for this, however I'm a little stuck on the naming. The handler is called `identitiesAuthenticatedGet`, so some potential options for the command are:
- `identities authenticated` - potentially too similar to `identities authenticate`?
- `identities linked`
- `identities connected` - might be confised with `identities search`
- `identities authorised`/`identities authorized`
- `identities authenticate --list` - make it a flag on the existing `identities authenticate` command
- Scrap the command and handler all together and just use `identities get [identityId]` to display own gestalt - requires user to know at least one of their authenticated identities + requires https://github.com/MatrixAI/js-polykey/issues/319

Reply:
Something like this:

```
pk identities tokens
```

The `tokens` is what they are.

Reply:
The `token`s stored in `IdentitiesManager` (what is retrieved by `identitiesTokenGet`) are actually `TokenData`s:
```ts
type TokenData = {
  accessToken: string;
  refreshToken?: string;
  accessTokenExpiresIn?: number;
  refreshTokenExpiresIn?: number;
};
```
I was also incorrect in my previous assumption that `identitiesTokenGet` was getting all tokens - you give it a `ProviderId` and an `IdentityId` and it returns the token for that identity id. Repurposing this handler for retrieving authenticated identity ids would also mean that `identitiesTokenPut` and `identitiesTokenDelete` would no longer have a corresponding 'getter'.

Is there a reason a user would need to be able to manipulate tokens at such a low level though? None of these handlers are exposed to the CLI anyway so maybe they should just all be removed? Then we can use the word 'token' to mean 'authenticated identity id' throughout the CLI without worrying about it being confused with `TokenData`?

Reply:
I forgot about these other commands. If so, then you could keep it as `identitiesTokenGet` and `pk identities token` command instead.

We'll look into the overall CLI later, let's keep this simple for now and just token manipulation is fine. And it's a CLI, the GUI may further abstract.

Reply:
Since this PR doesn't really deal with token manipulation, I think writing CLI commands for the token manipulation handlers can probably just wait until the general CLI review.

If we still want a command for listing the identities we currently have authenticated, then it sounds like this would be separate from the token manipulation commands. But in that case we're back to the original question of what should the command that does this be called. Again, this functionality isn't necessarily required in order to finish off this PR, and it can always be addressed later.

Reply:
Going back to this
> BTW, I remember now that we made a decision to group all gestalt-related functionality within the identities subcommand. So there is no pk gestalts subcommand. This was a UX decision so that users don't have to know what "gestalt" means. We can reverse this decision, but for now most of what we want to do will be identity focused.

This PR probably isn't the place for it, but I think separating the identities and gestalts commands is something we will eventually need to do, and I think the naming difficulties we're having here are a good example of why. We already have `identities list` and `identities get` commands, either of these would be good options for a command that lists/gets your own auth identities, however they're already being used for gestalts-related subcommands. 

Again, this sort of thing is outside the scope of this PR, so I'm leaning towards creating a separate issue for reviewing the identities CLI specifically (including everything that has been brought up in this thread) and linking it to the general CLI review issue.

Reply:
This will be resolved in a new issue: https://github.com/MatrixAI/js-polykey/issues/334